### PR TITLE
F/user discard

### PR DESCRIPTION
**Pull request by aaguilarg**
### Contexto
Cuando se eliminan usuarios, se elimina toda su historia. La idea es mantenerla.​

### Qué se esta haciendo
- Se agrega gema discard para esta logica
- En endpoints de usuarios se exponen solo los kept
- Para team members lo mismo, ademas de que el sorteo se hace solo con kept (se saca de la cola los descartados)
​
#### En particular hay que revisar
​

-----------

### CODE REVIEW

File: spec/controllers/api/bearers/v1/users_controller_spec.rb [hunk 1]
```diff
@@ -11,7 +11,7 @@ def attach_api_key(api_key)
 
   before { attach_api_key(organization_client) }
 
-  describe 'GET index' do
+ describe 'GET index' do
```
Comment (monkeyci) [inline spec/controllers/api/bearers/v1/users_controller_spec.rb:14]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Correctable] Layout/IndentationConsistency: Inconsistent indentation detected.

File: spec/controllers/api/v1/users_controller_spec.rb [hunk 2]
```diff
@@ -7,6 +7,7 @@
   describe 'GET #index' do
     let(:users_list_size) { 2 }
     let!(:users_list) { create_list(:user, users_list_size, organization: user.organization) }
+    let!(:discarded_user) { create(:user, organization: user.organization, discarded_at: Time.current) }
```
Comment (monkeyci) [inline spec/controllers/api/v1/users_controller_spec.rb:10]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Correctable] Layout/LineLength: Line is too long. [104/100]

File: spec/engines/tasks/extensions/tasks/team_member_observer_ext_spec.rb [hunk 3]
```diff
@@ -35,6 +35,20 @@ def trigger(type, event)
     end
   end
 
+  fcontext 'when saved' do
```
Comment (monkeyci) [inline spec/engines/tasks/extensions/tasks/team_member_observer_ext_spec.rb:38]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Correctable] RSpec/Focus: Focused spec found.

File: app/controllers/api/bearers/v1/users_controller.rb [hunk 4]
```diff
@@ -1,6 +1,6 @@
 class Api::Bearers::V1::UsersController < Api::Bearers::V1::BaseController
   def index
-    respond_with current_organization.users
+    respond_with current_organization.users.kept
```
Comment (monkeyci) [inline app/controllers/api/bearers/v1/users_controller.rb:3]: <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Not covered lines: 3

File: app/controllers/api/v1/team_members_controller.rb [hunk 5]
```diff
@@ -1,6 +1,6 @@
 class Api::V1::TeamMembersController < Api::V1::BaseController
   def index
-    respond_with policy_scope(TeamMember.includes(:user).where(team_id: params[:team_id]))
+    respond_with policy_scope(TeamMember.kept.includes(:user).where(team_id: params[:team_id]))
```
Comment (monkeyci) [inline app/controllers/api/v1/team_members_controller.rb:3]: <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Not covered lines: 3

File: app/controllers/api/v1/users_controller.rb [hunk 6]
```diff
@@ -1,6 +1,6 @@
 class Api::V1::UsersController < Api::V1::BaseController
   def index
-    respond_with current_user.organization.users.includes(:invitation), with_invitations: true
+    respond_with current_user.organization.users.kept.includes(:invitation), with_invitations: true
```
Comment (monkeyci) [inline app/controllers/api/v1/users_controller.rb:3]: <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Not covered lines: 3

File: app/models/user.rb [hunk 7]
```diff
@@ -15,6 +17,14 @@ class User < ApplicationRecord
 
   validates :first_name, :last_name, :email, presence: true
 
+  after_discard do
+    team_members.discard_all
```
Comment (monkeyci) [inline app/models/user.rb:21]: <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Not covered lines: 21

File: app/models/user.rb [hunk 8]
```diff
@@ -15,6 +17,14 @@ class User < ApplicationRecord
...
+  end
+
+  def active_for_authentication?
+    super && !discarded?
```
Comment (monkeyci) [inline app/models/user.rb:25]: <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Not covered lines: 25

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -6,7 +6,7 @@ def profile
   end
 
   def index
-    @users = current_user.organization.users
+    @users = current_user.organization.users.kept
```
Comment (nicolasguridi) [inline app/controllers/users_controller.rb:9]: faltó el test de esto o no?

**Review by nicolasguridi — APPROVED** [review]
te dejé un pequeño comentario 👌 ojo con los muchos comentarios del mono eso sí jajaja habían algunos razonables

**Pull request merged**